src/blk/kernel/KernelDevice.cc

```diff
@@ -878,9 +878,10 @@ int KernelDevice::write(
     return 0;
   }
 
-  if ((!buffered || bl.get_num_buffers() >= IOV_MAX) &&
-      bl.rebuild_aligned_size_and_memory(block_size, block_size, IOV_MAX)) {
-    dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
+  if ((!buffered || bl.get_num_buffers() >= IOV_MAX)) {
+    if (unsigned memcpy = bl.rebuild_aligned_size_and_memory(block_size, block_size, IOV_MAX) > 0) {
```

Comment:
this is buggy. the assignment (`=`) operator has lower precedence than that of the less than operator (`>`), so it's equivalent to

```c++
bool greater_than_zero = bl.rebuild_aligned_size_and_memory(block_size, block_size, IOV_MAX) > 0;
unsigned memcpy = greater_than_zero;
```

hence `memcpy` is always `1` here. the same applies to other places in this change.

and i'd avoid using the name of a well known libc function for naming variables.